cpp/src/arrow/compute/kernels/vector_dictionary.cc

```diff
@@ -0,0 +1,210 @@
+    if (indice->length() == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indice, empty_dict);
+    }
+    const CType* indices_data =
+        reinterpret_cast<const CType*>(indice->data()->buffers[1]->data());
+    int64_t offset = indice->data()->offset;
+
+    // check whether the input is compacted
+    std::vector<bool> dict_used(dict->length(), false);
+    int64_t dict_used_count = 0;
+    for (int64_t i = 0; i < indice->length(); i++) {
+      if (indice->IsNull(i)) {
+        continue;
+      }
+
+      CType cur_indice = indices_data[i + offset];
+      if (!dict_used[cur_indice]) {
+        dict_used[cur_indice] = true;
+        dict_used_count++;
+
+        if (dict_used_count == dict->length()) {  // input is already compacted
+          return dict_array;
+        }
+      }
+    }
+
+    // dictionary compaction
+    if (dict_used_count == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indice, empty_dict);
+    }
+    std::vector<CType> dict_indice;
+    bool need_change_indice = false;
+    CType len = static_cast<CType>(dict->length());
+    for (CType i = 0; i < len; i++) {
+      if (dict_used[i]) {
+        dict_indice.push_back(i);
+      } else if (i + 1 < len && dict_used[i + 1]) {
+        need_change_indice = true;
+      }
+    }
+    BuilderType dict_indice_builder;
+    ARROW_RETURN_NOT_OK(dict_indice_builder.AppendValues(dict_indice));
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::Array> compacted_dict_indices,
+                          dict_indice_builder.Finish());
+    ARROW_ASSIGN_OR_RAISE(
+        auto compacted_dict_res,
+        Take(dict, compacted_dict_indices, TakeOptions::NoBoundsCheck(), ctx));
+    std::shared_ptr<arrow::Array> compacted_dict = compacted_dict_res.make_array();
+
+    // indice changes
+    if (!need_change_indice) {
+      return DictionaryArray::FromArrays(dict_array->type(), indice, compacted_dict);
+    }
+    std::vector<CType> indice_minus_number(dict->length(), 0);
+    if (!dict_used[0]) {
+      indice_minus_number[0] = 1;
+    }
+    for (int64_t i = 1; i < dict->length(); i++) {
+      indice_minus_number[i] = indice_minus_number[i - 1];
+      if (!dict_used[i]) {
+        indice_minus_number[i] = indice_minus_number[i] + 1;
+      }
+    }
+
+    std::vector<CType> raw_changed_indice(indice->length(), 0);
+    std::vector<bool> is_valid(indice->length(), true);
+    for (int64_t i = 0; i < indice->length(); i++) {
+      if (indice->IsNull(i)) {
+        is_valid[i] = false;
+      } else {
+        CType cur_indice = indices_data[i + offset];
+        raw_changed_indice[i] = cur_indice - indice_minus_number[cur_indice];
+      }
+    }
+    BuilderType indice_builder;
```

Comment:
Please use the context's memory pool; it may have overridden the default
```suggestion
    BuilderType indices_builder(ctx->memory_pool());
```